core/sql/sqlcomp/CmpSeabaseDDLroutine.cpp

```diff
@@ -563,32 +795,332 @@ void CmpSeabaseDDL::dropSeabaseLibrary(StmtDDLDropLibrary * dropLibraryNode,
+					 NAString &currCatName, 
+					 NAString &currSchName)
+{
+  Lng32 cliRC;
+  Lng32 retcode;
+  
+  NAString libraryName = alterLibraryNode->getLibraryName();
+  NAString libFileName = alterLibraryNode->getFilename();
+  
+  ComObjectName libName(libraryName, COM_TABLE_NAME);
+  ComAnsiNamePart currCatAnsiName(currCatName);
+  ComAnsiNamePart currSchAnsiName(currSchName);
+  libName.applyDefaults(currCatAnsiName, currSchAnsiName);
+  
+  NAString catalogNamePart = libName.getCatalogNamePartAsAnsiString();
+  NAString schemaNamePart = libName.getSchemaNamePartAsAnsiString(TRUE);
+  NAString libNamePart = libName.getObjectNamePartAsAnsiString(TRUE);
+  const NAString extLibName = libName.getExternalName(TRUE);
+  
+  ExeCliInterface cliInterface(STMTHEAP, 0, NULL,
+			       CmpCommon::context()->sqlSession()->getParentQid());
+  
+  retcode = existsInSeabaseMDTable(&cliInterface, 
+				   catalogNamePart, schemaNamePart, libNamePart,
+				   COM_LIBRARY_OBJECT, TRUE, FALSE);
+  if (retcode < 0)
+    {
+      processReturn();
+      return;
+    }
+  
+  if (retcode == 0) // does not exist
+    {
+      CmpCommon::diags()->clear();
+      *CmpCommon::diags() << DgSqlCode(-1389)
+			  << DgString0(extLibName);
+      processReturn();
+      return;
+    }
+  
+  // strip blank spaces
+  libFileName = libFileName.strip(NAString::both, ' ');
+  if (validateLibraryFileExists(libFileName, FALSE))
+    {
+      processReturn();
+      return;
+    }
+  
+  Int32 objectOwnerID = 0;
+  Int32 schemaOwnerID = 0;
+  Int64 objectFlags = 0;
+  Int64 libUID = getObjectInfo(&cliInterface,
+			       catalogNamePart.data(), schemaNamePart.data(),
+			       libNamePart.data(), COM_LIBRARY_OBJECT,
+			       objectOwnerID,schemaOwnerID,objectFlags);
+  
+  // Check for error getting metadata information
+  if (libUID == -1 || objectOwnerID == 0)
+    {
+      if (CmpCommon::diags()->getNumber(DgSqlCode::ERROR_) == 0)
+	SEABASEDDL_INTERNAL_ERROR("getting object UID and owner for alter library");
+      processReturn();
+      return;
+    }
+  
+  // Verify that the current user has authority to perform operation
+  if (!isDDLOperationAuthorized(SQLOperation::ALTER_LIBRARY,
+				objectOwnerID,schemaOwnerID))
+    {
+      *CmpCommon::diags() << DgSqlCode(-CAT_NOT_AUTHORIZED);
+      processReturn();
+      return;
+    }
+  
+  Int64 redefTime = NA_JulianTimestamp();
+    size_t lastSlash = libFileName.last('/');
+  NAString libNameNoPath;
+  if (lastSlash != NA_NPOS)
+    libNameNoPath = libFileName(lastSlash+1, libFileName.length()-lastSlash-1);
+  else
+    {
+      *CmpCommon::diags() << DgSqlCode(-1382)
+                        << DgString0(libFileName);
+      processReturn();
+      return;
+      
+    }
+  char buf[2048]; // filename max length is 512. Additional bytes for long
+  // library names.
+  str_sprintf(buf, "update %s.\"%s\".%s set library_filename = '%s' , library_storage = filetolob('%s') where library_uid = %ld",
```

Comment:
This brings up a question: It looks like we overwrite the LIBRARIES row when we alter a library. That means that if the old library is not already cached, it will no longer be available. Initially, I thought that we might keep the old version available in the metadata, but maybe that's too complicated. Do you think we could run into an error in the following situation:

```
Process a: prepare s from call sp1();
Process b: alter library lib1 .... ; -- lib1 is the library for sp1
Process a: execute s;
```

Now, if we assume that the jar file for lib1 is not yet in the cache, executing s will try to extract it, but it won't find it in the metadata anymore - assuming that updating the table will invalidate the LOB handle we used for the previous version?

Reply:
I'v eadded QI support for alter library. So when an alter library is done, the objUid of the library is added to the list of invalidationKeys. So the prepare by process 'a' will get invalidated . So we avoid the issue of a stale library being executed .

Reply:
Couldn't there still be a short window where one query is executing and another process upgrades the library? If so, that's probably ok, mostly wondering whether I understand this correctly.

Reply:
Maybe a small window where the alter of the library has finished and another process has started executing but  not yet extracted the library to cache.  Not sure what we can do about that. If the library is already cached it'll just use the old copy.  I think we'll need to simulate this and address it separately.. If needed we may need to maintain the old version of the library.